core/src/main/java/org/apache/shiro/realm/text/TextConfigurationRealm.java

```diff
@@ -211,11 +212,14 @@ protected void processUserDefinitions(Map<String, String> userDefs) {
 
     protected static Set<String> toLines(String s) {
         LinkedHashSet<String> set = new LinkedHashSet<String>();
-        Scanner scanner = new Scanner(s);
-        while (scanner.hasNextLine()) {
-            set.add(scanner.nextLine());
-        }
-        return set;
+		try (Scanner scanner = new Scanner(s)) {
+			while (scanner.hasNextLine()) {
+				set.add(scanner.nextLine());
+			}
+		} catch (NoSuchElementException e) {
+			e.printStackTrace();
+		} 
+		return set;
```

Comment:
The rest of the file is formatted with spaces, this blog looks like tabs? can you fix that?

Instead of `e.printStackTrace()` use a logger.

More importantly though, how was this the exception triggered?  I'm guessing if `hasNextLine()` returns true, and then `.nextLine()` throws a `NoSuchElementException`, that's not an error you would want to recover from, as the stream would have been corrupted?  And in this case potentially worse because we are operating against a String in memory, which means things have gone severely wrong?

I could be miss understanding something here though, I only took a quick look.

Reply:
Link to Apache's SonarCloud: [the bug](https://sonarcloud.io/project/issues?id=apache_shiro&languages=java&open=AXTvO9kn3S_U1xqKcObL&resolved=false&severities=BLOCKER&types=BUG).

Reply:
> how was this the exception triggered? 

> Link to Apache's SonarCloud: the bug.

yields: `Use try-with-resources or close this "Scanner" in a "finally" clause.`

That is, you would not want to catch the exception. Just the try would be fine in this case, as this is only about closing the scanner. But as we are reading a String, this is a noop anyway:

https://github.com/openjdk/jdk/blob/d7268fa3a68c2356548651a27b307d7f7158e700/src/java.base/share/classes/java/util/Scanner.java#L1171-L1175

So in this case, while Sonar is right in the sense of "good habit" (as the source could change): My -1 on this as we are not "fixing" anything or making anything better. Closing the scanner is fine for me, but there was no underlying bug and the API would also never be able to change to something `Closeable`.